PopED + babelmixr2 for windows

mattfidler opened the pull request:
Hi @andrewhooker 

This allows windows based `PopED` models from babelmixr2 to be run in parallel without modifying any user code.  This mean the same user code works on windows and linux/mac

mattfidler commented:
In general it saves the `babelmixr2` environment inside of the `poped.db` setup from `babelmixr2`.  With the windows parallelization, this environment is sent to every R process node to load the DLLs.  This will allow everything to be contained inside of `babelmixr2` in case things like symbolic sensitivity gradient models are added to speed up `PopED` and `rxode2`.  Right now there are 2 models/dlls generated for each poped database.

mattfidler commented:
Had to add `babelmixr2` development stack so these check correctly.  After adding this to CRAN you can take out the `babelmixr2` stack and can simply use the CRAN version.

mattfidler committed:
PopED + babelmixr2 for windows

mattfidler committed:
Add babelmixr2 to suggests

mattfidler committed:
Add to news

mattfidler committed:
add babelmixr2 dev for check

mattfidler committed:
Revert "add babelmixr2 dev for check"

This reverts commit 4b136a024736ed46cdd34c1da82bd503473bfcb0.

mattfidler commented:
As `babelmixr2` was released on `CRAN` development libraries are no longer neeeded.

mattfidler commented:
The error in oldrel-1 is unrelated to this pull request.

andrewhooker committed:
Merge branch 'master' into babelmixr2-win

mattfidler commented:
Thanks Andy!

andrewhooker commented on R/start_parallel.R:
Why do we need to add 

```
start_parallel_env <- new.env(parent=emptyenv())
start_parallel_env$babelmixr2 <- NULL
```
 
outside of the function?

mattfidler replied:
This is so it can be assigned in a global like environment.

The actual function is

```r
parallel::clusterCall(cl, babelmixr2::.popedCluster, start_parallel_env$babelmixr2)
```

andrewhooker commented:
Hi Matt,

Thanks for all your work on this. I am wondering if there is any way to achieve the same result without having to add the following code snippet in, essentially, every PopED function?

`start_parallel_env$babelmixr2 <- poped.db$babelmixr2`

Could you supply a minimal working example that I could test a bit with? My thought is that it should be a bit more efficient to just pass `poped.db` to  `start.parallel()` 

Andy

mattfidler commented:
parallel::clusterCall(cl, babelmixr2::.popedCluster, start_parallel_env$babelmixr2)

> Hi Matt,
> 
> Thanks for all your work on this. I am wondering if there is any way to achieve the same result without having to add the following code snippet in, essentially, every PopED function?
> 
> `start_parallel_env$babelmixr2 <- poped.db$babelmixr2`
> 
> Could you supply a minimal working example that I could test a bit with? My thought is that it should be a bit more efficient to just pass `poped.db` to `start.parallel()`
> 
> Andy

I originally tried this, but it dropped the `$babelmixr2` in some functions.  Perhaps you know how to keep it from dropping.

mattfidler commented:
I simply used your examples and modified it.  

You could try the warfarin one:

https://github.com/nlmixr2/babelmixr2/blob/175ca39b033df3b5020b7769157c175721abf29f/inst/poped/ex.2.c.warfarin.ODE.compiled.babelmixr2.R

andrewhooker committed:
Simplify babelmixr2 Windows parallel support

start_parallel() now auto-detects poped.db$babelmixr2 from parent.frame(), the same mechanism already used for clusterExport(). This removes the per-function boilerplate and module-level global state.

andrewhooker commented:
I uppdated start_parallel() so that it now auto-detects poped.db$babelmixr2 from parent.frame(), the same mechanism already used for clusterExport(). This removes the per-function boilerplate and module-level global state. I think this solution should work. Can you test it?

mattfidler commented:
Hi Andy, 

Your modifications do not work.

With the simplest functions in PopED babelmixr2 I get:

<img width="1853" height="215" alt="image" src="https://github.com/user-attachments/assets/61bb4198-b23a-4ec2-9daa-46f0f6a4287f" />

This is for:

https://github.com/nlmixr2/babelmixr2/blob/main/inst/poped/ex.1.b.PK.1.comp.oral.md.re-parameterize.babelmixr2.R

mattfidler commented:
Reading the code, it is a bit fragile; it assumes you are using the name `poped.db` for the PopED database which is definitely not required.

andrewhooker committed:
updated roxygen

andrewhooker committed:
Updates to allow for babelmixr

andrewhooker committed:
Merge branch 'master' into babelmixr2-win

andrewhooker commented:
Hi Matt,

I have tried to update the fix to make it more robust. I had a hard time installing babelmixr2 on my parallels version of windows, so I would be happy if you could test it. I notice that with 

https://github.com/nlmixr2/babelmixr2/blob/main/inst/poped/ex.1.b.PK.1.comp.oral.md.re-parameterize.babelmixr2.R

that using babelmixr2 instead of the original example causes a slowdown of many orders of magnitude. Not quite sure why.

andrewhooker committed:
Merge branch 'master' into babelmixr2-win

mattfidler committed:
Fix babelmixr2 requirement

mattfidler commented:
Just tested, it still doens't work:

```r
library(babelmixr2)

library(PopED)

##-- Model: One comp first order absorption
## -- Analytic solution for both mutiple and single dosing

f <- function() {
  ini({
    tV <- 72.8
    tKa <- 0.25
    tCl <- 3.75
    tF <- fix(0.9)

    eta.v ~ 0.09
    eta.ka ~ 0.09
    eta.cl ~0.25^2

    prop.sd <- fix(sqrt(0.04))
    add.sd <- fix(sqrt(5e-6))

  })
  model({
    V<-tV*exp(eta.v)
    KA<-tKa*exp(eta.ka)
    CL<-tCl*exp(eta.cl)
    Favail <- tF
    N <-  floor(time/TAU)+1
    y <- (DOSE*Favail/V)*(KA/(KA - CL/V)) *
      (exp(-CL/V * (time - (N - 1) * TAU)) *
         (1 - exp(-N * CL/V * TAU))/(1 - exp(-CL/V * TAU)) -
         exp(-KA * (time - (N - 1) * TAU)) * (1 - exp(-N * KA * TAU))/(1 - exp(-KA * TAU)))

    y ~ prop(prop.sd) + add(add.sd)
  })
}

e <- et(list(c(0, 1, 10),
             c(0, 2, 10),
             c(0, 8, 10),
             c(240, 240, 248),
             c(240, 245, 248))) %>%
  as.data.frame()



babel.db <- nlmixr2(f, e, "poped",
                        popedControl(groupsize=20,
                                     bUseGrouped_xt=TRUE,
                                     a=list(c(DOSE=20,TAU=24),
                                            c(DOSE=40, TAU=24)),
                                     maxa=c(DOSE=200,TAU=24),
                                     mina=c(DOSE=0,TAU=24)))

##  create plot of model without variability
plot_model_prediction(babel.db, model_num_points = 300)

##  create plot of model with variability
plot_model_prediction(babel.db, IPRED=T, DV=T, separate.groups=T, model_num_points = 300)

## evaluate initial design
## From original
## $rse
##        V        KA        CL       d_V      d_KA      d_CL
## 8.215338 10.090955  4.400304 39.833230 60.089601 27.391518

evaluate_design(babel.db)

## original: > shrinkage(poped.db)
## # A tibble: 9 × 5
##     d_V  d_KA   d_CL type       group
##   <dbl> <dbl>  <dbl> <chr>      <chr>
## 1 0.364 0.578 0.184  shrink_var all_groups
## 2 0.364 0.579 0.184  shrink_var grp_1
## 3 0.363 0.577 0.183  shrink_var grp_2
## 4 0.202 0.350 0.0965 shrink_sd  all_groups
## 5 0.202 0.351 0.0967 shrink_sd  grp_1
## 6 0.202 0.350 0.0963 shrink_sd  grp_2
## 7 0.181 0.228 0.107  se         all_groups
## 8 0.181 0.228 0.107  se         grp_1
## 9 0.181 0.228 0.107  se         grp_2
shrinkage(babel.db)

# Optimization of sample times
# Note: The parallel option does not work well with Windows machines at this moment.
# Please set parallel = FALSE if you are working on a Windows machine
output <- poped_optim(babel.db, opt_xt =TRUE, parallel=TRUE)

# Evaluate optimization results
summary(output)

## From original
# V        KA        CL       d_V      d_KA      d_CL
# 6.281944  7.726279  4.295908 32.416232 49.062880 26.363021
get_rse(output$FIM,output$poped.db)

plot_model_prediction(output$poped.db)

# Optimization of sample times and doses
# Note: The parallel option does not work well with Windows machines at this moment.
# Please set parallel = FALSE if you are working on a Windows machine
output_2 <- poped_optim(output$poped.db, opt_xt =TRUE, opt_a = TRUE, parallel = TRUE)
```

Gives

```
************* Iteration 1  for all optimization methods***********************

*******************************************
Running Adaptive Random Search Optimization
*******************************************
Error in `start_parallel()`:
! object 'babelmixr2_model' not found
Hide Traceback
Fix
Explain
    ▆
 1. └─PopED::poped_optim(babel.db, opt_xt = TRUE, parallel = TRUE)
 2.   ├─base::do.call(poped_optim_3, c(new.arg.list, dot_vals))
 3.   └─PopED (local) `<fn>`(...)
 4.     ├─base::do.call(...)
 5.     └─PopED (local) `<fn>`(...)
 6.       └─PopED::start_parallel(...)
 ```

andrewhooker committed:
Updated roxygen

andrewhooker committed:
fix spelling error

andrewhooker committed:
update documentation

andrewhooker committed:
merging with master branch

Merge branch 'master' into pr-79

# Conflicts:
#	.Rbuildignore
#	.gitignore
#	DESCRIPTION
#	NEWS.md
#	inst/examples/ex.1.b.PK.1.comp.oral.md.re-parameterize.R

andrewhooker committed:
merge updates

andrewhooker committed:
Merge branch 'master' into babelmixr2-win
